wolfcrypt/src/port/nxp/dcp_port.c

```diff
@@ -0,0 +1,515 @@
+    }
+
+    else if (keySelect == kDCP_OCOTPKeyLow)
+    {
+        IOMUXC_GPR->GPR3 &= ~(1 << IOMUXC_GPR_GPR3_DCP_KEY_SEL_SHIFT);
+        IOMUXC_GPR->GPR10 |= (1 << IOMUXC_GPR_GPR10_DCPKEY_OCOTP_OR_KEYMUX_SHIFT);
+    }
+
+    else if (keySelect == kDCP_OCOTPKeyHigh)
+    {
+        IOMUXC_GPR->GPR3 |= (1 << IOMUXC_GPR_GPR3_DCP_KEY_SEL_SHIFT);
+        IOMUXC_GPR->GPR10 |= (1 << IOMUXC_GPR_GPR10_DCPKEY_OCOTP_OR_KEYMUX_SHIFT);
+    }
+
+    else
+    {
+        retval = kStatus_InvalidArgument;
+    }
+    dcp_unlock();
+    return retval;
+}
+#endif
+
+static const int dcp_channels[4] = {
+    kDCP_Channel0,
+    kDCP_Channel1,
+    kDCP_Channel2,
+    kDCP_Channel3
+};
+static int dcp_status[4] = {0, 0, 0, 0};
+
+static int dcp_get_channel(void)
+{
+#ifdef SINGLE_THREADED
+    return dcp_channels[0];
+#else
+    int i;
+    for (i = 0; i < 4; i++) {
+        if (dcp_status[i] == 0) {
+            dcp_status[i]++;
+            return dcp_channels[i];
+        }
+    }
+    return 0;
+#endif
+}
+
+
+static int dcp_init(void)
+{
+    int ch = dcp_get_channel();
+    if (!ch) {
+        return 0;
+    }
+    if (!dcp_is_initialized) {
+        dcp_config_t dcpConfig;
+#ifndef SINGLE_THREADED
+        /* Initialize and lock mutex */
+        wc_InitMutex(&dcp_mutex);
+        dcp_lock();
+#endif
+        DCP_GetDefaultConfig(&dcpConfig);
+        /* Initialize DCP */
+        /* Reset and initialize DCP */
+        DCP_Init(DCP, &dcpConfig);
+#if DCP_USE_OTP_KEY
+        /* Set OTP key type in IOMUX registers before initializing DCP. */
+        /* Software reset of DCP must be issued after changing the OTP key type. */
+        DCP_OTPKeySelect(kDCP_OTPMKKeyLow);
+#endif
+        dcp_is_initialized++;
+        /* Release mutex */
+        dcp_unlock();
+    }
+    return ch;
+}
+
+static void dcp_fini(int ch)
+{
+#ifdef SINGLE_THREADED
+    int i;
+    for (i = 0; i < 4; i++) {
+        if (ch == dcp_channels[i])
+            dcp_status[i] = 0;
+    }
+#endif
+}
+
+
+#ifndef NO_AES
+int DCPAesInit(Aes *aes)
+{
+    int ch;
+    if (!aes)
+        return BAD_FUNC_ARG;
+    ch = dcp_init();
+    if (ch == 0)
+        return WC_PENDING_E;
+    dcp_lock();
+    XMEMSET(&aes->handle, 0, sizeof(aes->handle));
+    aes->handle.channel    = ch;
+    aes->handle.swapConfig = kDCP_NoSwap;
+    dcp_unlock();
+    return 0;
+}
+
+int  DCPAesSetKey(Aes* aes, const byte* key, word32 len, const byte* iv,
+                          int dir)
+{
+
+#if DCP_USE_OTP_KEY
+#warning Please update cipherAes128 variables to match expected AES ciphertext for your OTP key.
+#endif
+    status_t status;
+    if (!aes || !key)
+        return BAD_FUNC_ARG;
+
+    if (len != 16)
+        return BAD_FUNC_ARG;
+
+    if (aes->handle.channel == 0)
+        return BAD_FUNC_ARG;
+
+    dcp_lock();
+#if DCP_USE_OTP_KEY
+    aes->handle.keySlot = kDCP_OtpKey;
+#else
+    aes->handle.keySlot = kDCP_KeySlot0;
```

Comment:
Should we add "dcp_get_keyslot" function to use 4 key slots parallelly?

Reply:
I agree. I have added a per-channel key slot assignment.